bftengine/src/bftengine/CryptoManager.cpp

```diff
@@ -0,0 +1,245 @@
+  std::lock_guard<std::mutex> guard(mutex_);
+  auto riter = checkpointToSystem().rbegin();
+  std::array<std::pair<SeqNum, std::shared_ptr<EdDSAMultisigVerifier>>, 2> result;
+  result[0] = {
+      riter->first,
+      std::reinterpret_pointer_cast<EdDSAMultisigVerifier>(riter->second->thresholdVerifierForOptimisticCommit_)};
+  riter++;
+  if (riter != checkpointToSystem().rend() && riter->second != nullptr) {
+    result[1] = {
+        riter->first,
+        std::reinterpret_pointer_cast<EdDSAMultisigVerifier>(riter->second->thresholdVerifierForOptimisticCommit_)};
+  }
+  return result;
+}
+
+std::array<std::shared_ptr<EdDSAMultisigSigner>, 2> CryptoManager::getLatestSigners() const {
+  std::lock_guard<std::mutex> guard(mutex_);
+  auto riter = checkpointToSystem().rbegin();
+  std::array<std::shared_ptr<EdDSAMultisigSigner>, 2> result;
+  result[0] = std::reinterpret_pointer_cast<EdDSAMultisigSigner>(riter->second->thresholdSigner_);
+  ++riter;
+  if (riter != checkpointToSystem().rend() && riter->second != nullptr) {
+    result[1] = std::reinterpret_pointer_cast<EdDSAMultisigSigner>(riter->second->thresholdSigner_);
+  }
+  return result;
+}
+
+CryptoManager::CryptoSystemWrapper::CryptoSystemWrapper(std::unique_ptr<Cryptosystem>&& cs)
+    : cryptosys_(std::move(cs)) {}
+
+void CryptoManager::CryptoSystemWrapper::init() {
+  std::uint16_t f{ReplicaConfig::instance().getfVal()};
+  std::uint16_t c{ReplicaConfig::instance().getcVal()};
+  std::uint16_t numSigners{ReplicaConfig::instance().getnumReplicas()};
+  thresholdSigner_.reset(cryptosys_->createThresholdSigner());
+  thresholdVerifierForSlowPathCommit_.reset(cryptosys_->createThresholdVerifier(f * 2 + c + 1));
+  thresholdVerifierForCommit_.reset(cryptosys_->createThresholdVerifier(f * 3 + c + 1));
+  thresholdVerifierForOptimisticCommit_.reset(cryptosys_->createThresholdVerifier(numSigners));
+}
+
+CheckpointNum CryptoManager::getCheckpointOfCryptosystemForSeq(const SeqNum sn) const {
+  // find last chckp that is less than a chckp of a given sn
+  const uint64_t checkpointUpperBound = sn == 0 ? 0 : ((sn - 1) / checkpointWindowSize);
+  for (auto riter = checkpointToSystem().rbegin(); riter != checkpointToSystem().rend(); ++riter) {
+    auto& [checkpointCandidate, cryptosystem] = *riter;
+    if (checkpointCandidate <= checkpointUpperBound) {
+      LOG_DEBUG(logger(),
+                "Found cryptosystem for " << KVLOG(sn,
+                                                   checkpointUpperBound,
+                                                   checkpointCandidate,
+                                                   cryptosystem,
+                                                   checkpointToSystem().size(),
+                                                   checkpointWindowSize));
+      return checkpointCandidate;
+    }
+  }
+
+  // Replicas might encounter old messages, crashing is not
```

Comment:
Assuming we couldn't find a cryptosystem that matches the given sequence number (for example, we got a message from a lagged replica).
Why does it make sense to take the newest one (rbegin) on top of the oldest one?

Reply:
Both will probably not be able to validate, the code for getting the most recent one was the simplest.
I don't mind changing it either way.

Reply:
up to you :) 
I only wanted to understand